File: chroma-league/src/main/java/com/bonepl/chromaleague/hud/colors/GoldColor.java
package com.bonepl.chromaleague.hud.colors;

import com.bonepl.razersdk.color.Color;
import com.bonepl.razersdk.color.MultiTransitionColor;
import com.bonepl.razersdk.color.StaticColor;

import java.security.SecureRandom;

public class GoldColor implements Color {
    private static final StaticColor DARKER_YELLOW = new StaticColor(60, 60, 0);
    private static final SecureRandom SECURE_RANDOM = new SecureRandom();
    private static final float BLINK_CHANCE = 0.02f;
    private int blinkingCountdown;
    MultiTransitionColor blinkColor = new MultiTransitionColor.Builder(StaticColor.YELLOW)
            .addTransition(StaticColor.WHITE, 3)
            .looped(3)
            .build();
    MultiTransitionColor goldColor = new MultiTransitionColor.Builder(StaticColor.YELLOW)
            .addTransition(DARKER_YELLOW, 20)
            .looped(20)
            .build();

    @Override
    public StaticColor getColor() {
        if (blinkingCountdown == 0) {
            if (SECURE_RANDOM.nextFloat() <= BLINK_CHANCE) {
                blinkingCountdown = 4;
            }
            return goldColor.getColor();
        } else {
            blinkingCountdown--;
            return blinkColor.getColor();
        }
    }
}

File: src/hud/colors/GoldColor.ts
import {Color} from '../../razer-sdk/color/Color.js';
import {MultiTransitionColor} from '../../razer-sdk/color/MultiTransitionColor.js';
import {StaticColor} from '../../razer-sdk/color/StaticColor.js';

const DARKER_YELLOW = new StaticColor(60, 60, 0);
const BLINK_CHANCE = 0.01;

export class GoldColor implements Color {
    private blinkingCountdown = 0;
    private blinkColor: MultiTransitionColor | null = null;
    private readonly goldColor = new MultiTransitionColor.Builder(StaticColor.YELLOW)
        .addTransition(DARKER_YELLOW, 20)
        .looped(20)
        .build();

    getColor(): StaticColor {
        if (this.blinkingCountdown === 0) {
            if (Math.random() <= BLINK_CHANCE) {
                this.blinkingCountdown = 4;
                this.blinkColor = this.createBlinkColor();
            }
            return this.goldColor.getColor();
        } else {
            this.blinkingCountdown--;
            return this.blinkColor!.getColor();
        }
    }

    private createBlinkColor(): MultiTransitionColor {
        return new MultiTransitionColor.Builder(this.goldColor.getColor())
            .addTransition(StaticColor.WHITE, 3)
            .looped(3)
            .build();
    }
}
